File: impl/ts/src/models-config.ts
// Runtime-editable model fallback config. Users can add models from multiple
// providers (OpenRouter, z.ai, or any OpenAI-compatible base URL) and set the
// priority order. Persisted to a JSON file next to .env so it survives restart;
// falls back to the OPENROUTER_MODEL env list on first run.

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

// A single model entry in the ordered fallback chain.
export interface ModelEntry {
  // The model id sent to the provider, e.g. "google/gemma-4-31b-it:free" or
  // "glm-4.6". For the built-in OpenRouter provider, this is the bare id; for
  // other providers it is whatever that provider expects.
  model: string;
  // Provider routing key: "openrouter" (default), "zai", or "custom".
  provider: string;
  // Only for provider === "custom": the OpenAI-compatible base URL + the env
  // var name holding the key (so keys never live in this JSON).
  baseUrl?: string;
  keyEnv?: string;
  // Whether this entry participates in the fallback chain.
  enabled: boolean;
}

function configPath(): string {
  // impl/ts/src/models-config.ts -> impl/ts/models.json
  const here = path.dirname(fileURLToPath(import.meta.url));
  return path.resolve(here, "..", "models.json");
}

function envPath(): string {
  const here = path.dirname(fileURLToPath(import.meta.url));
  return path.resolve(here, "..", ".env");
}

// Parse the OPENROUTER_MODEL env list into entries (the bootstrap default).
function defaultEntries(): ModelEntry[] {
  let raw = process.env.OPENROUTER_MODEL ?? "";
  if (!raw) {
    try {
      const text = fs.readFileSync(envPath(), "utf-8");
      for (const line of text.split(/\r?\n/)) {
        const t = line.trim();
        if (t.startsWith("OPENROUTER_MODEL=")) {
          raw = t.slice("OPENROUTER_MODEL=".length).trim();
          break;
        }
      }
    } catch {
      /* no .env is fine */
    }
  }
  if (!raw) raw = "openrouter/owl-alpha";
  return raw
    .split(",")
    .map((m) => m.trim())
    .filter(Boolean)
    .map((m) => entryFromToken(m));
}

// Turn a chain token like "zai:glm-5.1" into a structured entry.
export function entryFromToken(token: string): ModelEntry {
  if (token.startsWith("zai:")) {
    return { model: token.slice(4), provider: "zai", enabled: true };
  }
  return { model: token, provider: "openrouter", enabled: true };
}

// Turn an entry back into a chain token the LLM client understands.
export function tokenFromEntry(e: ModelEntry): string {
  if (e.provider === "zai") return `zai:${e.model}`;
  // Custom OpenAI-compatible host: encode keyEnv + baseUrl + model so the LLM
  // client can route there. Format must match the client's resolve() parser.
  if (e.provider === "custom") return `custom:${e.keyEnv ?? ""}|${e.baseUrl ?? ""}|${e.model}`;
  return e.model;
}

let cache: ModelEntry[] | null = null;

export function getModels(): ModelEntry[] {
  if (cache) return cache;
  try {
    const data = JSON.parse(fs.readFileSync(configPath(), "utf-8")) as ModelEntry[];
    if (Array.isArray(data) && data.length > 0) {
      cache = data;
      return cache;
    }
  } catch {
    /* no config yet -> bootstrap from env */
  }
  cache = defaultEntries();
  return cache;
}

export function setModels(entries: ModelEntry[]): ModelEntry[] {
  cache = entries.filter((e) => e.model && e.model.trim());
  fs.writeFileSync(configPath(), JSON.stringify(cache, null, 2), "utf-8");
  return cache;
}

// The ordered list of enabled chain tokens, for the LLM client.
export function enabledChain(): string[] {
  return getModels()
    .filter((e) => e.enabled)
    .map(tokenFromEntry);
}

File: impl/go/internal/agent/models_config.go
// Runtime-editable model fallback config. Users can add models from multiple
// providers (OpenRouter, z.ai, or any OpenAI-compatible base URL) and set the
// priority order. Persisted to models.json next to .env so it survives restart;
// falls back to the OPENROUTER_MODEL env list on first run.
package agent

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"sync"
)

// ModelEntry is one model in the ordered fallback chain.
type ModelEntry struct {
	Model    string `json:"model"`
	Provider string `json:"provider"` // "openrouter" (default), "zai", or "custom"
	BaseURL  string `json:"baseUrl,omitempty"`
	KeyEnv   string `json:"keyEnv,omitempty"`
	Enabled  bool   `json:"enabled"`
}

var (
	cfgMu    sync.Mutex
	cfgCache []ModelEntry
)

func modelsConfigPath() string {
	exe, _ := os.Executable()
	// binary lives in impl/go/ -> models.json beside it; fall back to cwd.
	if exe != "" {
		return filepath.Join(filepath.Dir(exe), "models.json")
	}
	return "models.json"
}

// entryFromToken turns a chain token like "zai:glm-5.1" into a structured entry.
func entryFromToken(token string) ModelEntry {
	if strings.HasPrefix(token, "zai:") {
		return ModelEntry{Model: token[4:], Provider: "zai", Enabled: true}
	}
	return ModelEntry{Model: token, Provider: "openrouter", Enabled: true}
}

// tokenFromEntry turns an entry back into a chain token the client understands.
func tokenFromEntry(e ModelEntry) string {
	switch e.Provider {
	case "zai":
		return "zai:" + e.Model
	case "custom":
		return "custom:" + e.KeyEnv + "|" + e.BaseURL + "|" + e.Model
	default:
		return e.Model
	}
}

func defaultEntries() []ModelEntry {
	raw := loadDotenv()["OPENROUTER_MODEL"]
	if v := os.Getenv("OPENROUTER_MODEL"); v != "" {
		raw = v
	}
	if raw == "" {
		raw = "openrouter/owl-alpha"
	}
	var out []ModelEntry
	for _, m := range strings.Split(raw, ",") {
		if m = strings.TrimSpace(m); m != "" {
			out = append(out, entryFromToken(m))
		}
	}
	return out
}

// GetModels returns the configured model entries (cached, bootstrapped from env).
func GetModels() []ModelEntry {
	cfgMu.Lock()
	defer cfgMu.Unlock()
	if cfgCache != nil {
		return cfgCache
	}
	if data, err := os.ReadFile(modelsConfigPath()); err == nil {
		var entries []ModelEntry
		if json.Unmarshal(data, &entries) == nil && len(entries) > 0 {
			cfgCache = entries
			return cfgCache
		}
	}
	cfgCache = defaultEntries()
	return cfgCache
}

// SetModels persists a new model list and updates the cache.
func SetModels(entries []ModelEntry) []ModelEntry {
	cfgMu.Lock()
	defer cfgMu.Unlock()
	var clean []ModelEntry
	for _, e := range entries {
		if strings.TrimSpace(e.Model) != "" {
			if e.Provider == "" {
				e.Provider = "openrouter"
			}
			clean = append(clean, e)
		}
	}
	cfgCache = clean
	if data, err := json.MarshalIndent(clean, "", "  "); err == nil {
		_ = os.WriteFile(modelsConfigPath(), data, 0o644)
	}
	return clean
}

// enabledChain is the ordered list of enabled chain tokens, for the client.
func enabledChain() []string {
	var out []string
	for _, e := range GetModels() {
		if e.Enabled {
			out = append(out, tokenFromEntry(e))
		}
	}
	return out
}
